siriuspy/siriuspy/devices/injsys.py

```diff
@@ -771,3 +773,269 @@ def _command_target(self, cmdtype):
+        dpk_on = (self.pudpk.pwrstate == PowerSupplyPU.PWRSTATE.On) and \
+            (self.pudpk.pulse == PowerSupplyPU.PULSTATE.On)
+        nlk_on = (self.punlk.pwrstate == PowerSupplyPU.PWRSTATE.On) and \
+            (self.punlk.pulse == PowerSupplyPU.PULSTATE.On)
+        return dpk_evt_ok & dpk_dly_ok & dpk_kck_ok & dpk_on & nlk_on
+
+    def cmd_switch_to_optim(self):
+        """Switch to Optimization."""
+        self._abort.clear()
+        self._update_status('Switching to PU Optimization config...')
+        if not self._check_pu_interlocks():
+            return False
+        if self._abort.is_set():
+            self._clr_flag_abort()
+            return False
+
+        # if previously in on-axis, do delta angle x
+        if self.is_onaxis:
+            delta = self.posang.delta_angx - self.TS_POSANG_DEFDELTA
+            if not self._do_delta_posang(delta):
+                return False
+
+        if self._abort.is_set():
+            self._clr_flag_abort()
+            return False
+
+        # configure DpK trigger
+        if not self._config_dpk_trigger(delayraw=self.SI_DPKCKR_DLYR_OPT):
+            return False
+
+        # set DpK Kick
+        if not self._config_dpk_kick():
+            return False
+
+        # set pulsed magnet pwrstate and pulse
+        proced = (
+            (self.pudpk.cmd_turn_on, 'turn DpK on.'),
+            (self.punlk.cmd_turn_on, 'turn NLK on.'),
+            (self.pudpk.cmd_turn_on_pulse, 'turn DpK pulse on.'),
+            (self.punlk.cmd_turn_on_pulse, 'turn NLK pulse on.'),
+        )
+        if not self._do_procedures(proced):
+            return False
+
+        self._update_status('PU mode switched to Optimization!')
+        return True
+
+    @property
+    def is_onaxis(self):
+        """If configuration is OnAxis."""
+        if not self.connected:
+            return False
+        dpk_evt_ok = self.trigdpk.source == self.trignlk.source
+        dpk_dly_ok = self.trigdpk.delay_raw == \
+            self.trignlk.delay_raw + self.SI_DPKCKR_DLYR_ONAXINC
+        dpk_kck_ok = self.pudpk.strength == self.SI_DPKCKR_DEFKICK
+        dpk_on = (self.pudpk.pwrstate == PowerSupplyPU.PWRSTATE.On) and \
+            (self.pudpk.pulse == PowerSupplyPU.PULSTATE.On)
+        nlk_off = (self.punlk.pwrstate == PowerSupplyPU.PWRSTATE.Off) or \
+            (self.punlk.pulse == PowerSupplyPU.PULSTATE.Off)
+        return dpk_evt_ok & dpk_dly_ok & dpk_kck_ok & dpk_on & nlk_off
+
+    def cmd_switch_to_onaxis(self):
+        """Switch to OnAxis."""
+        self._abort.clear()
+        self._update_status('Switching to PU OnAxis config...')
+        if not self._check_pu_interlocks():
+            return False
+        if self._abort.is_set():
+            self._clr_flag_abort()
+            return False
+
+        # if not previously in on-axis, do delta angle x
+        if not self.is_onaxis:
+            delta = self.posang.delta_angx + self.TS_POSANG_DEFDELTA
```

Comment:
I'm not sure whether this is the right way to proceed, because inside `_do_delta_posang` you may reset the reference. In this case I think it is not appropriate to apply the old `delta_angx`, because it may have been incorporated to new reference.

Reply:
Thanks for noticing the bug! Fixed!